website/docs/error-reference.md

```diff
@@ -3897,6 +3897,27 @@ arr = T::Array[NilClass].new
 T.unsafe(arr).dig(0, 0)
 ```
 
+## 7045
+
+Sorbet sometimes assumes an expression has a certain type, even when it has no
+guarantee that it's correct in that assumption, because in over 99.9% of cases
+the assumption is correct and allows for omitting an explicit type annotation.
+
+It makes these unconfirmed assumptions early in the course of type checking, at
+a point where it hasn't collected enough information to know what the right type
+is, and records a note to itself to check those assumptions before finishing.
+
+When one of those assumptions is broken, it reports this error. In every case,
+the fix is to provide the correct, explicit annotation so that Sorbet does not
+have to assume anything.
+
+(Sorbet works this way in general for performance—Sorbet can be much faster if
+it doesn't have to go back and re-do a lot of work it has already done when it
+encounters a broken assumption. There is more information on this in
+[Why does Sorbet sometimes need type annotations?](why-type-annotations.md). A
+few type annotations can go a long way to ensuring that Sorbet can type check
+large codebases quickly.)
```

Comment:
This paragraph strikes me as kind of weird given the preceding paragraphs explaining that Sorbet just tried to assume something and it turned out it was mistaken in that assumption.

e.g. what is "this way" in the first sentence referring to?  (It reads to me like it's referencing "...so that Sorbet doesn't have to assume anything"...but that referencing doesn't make sense in the context of this error message because clearly Sorbet is already making assumptions.)  Or what is "it doesn't have to go back and re-do a lot of work it has already done when it encounters a broken assumption" trying to get at?  I guess both of these are referencing the design principle that Sorbet does everything in a single pass, no backtracking or fixed points required, but I'm not sure how useful that is as user documentation.

Maybe this whole paragraph would just be better as a quick sentence pointing people at `why-type-annotations.md`?  But that also seems strange, since `why-type-annotations.md` is explaining, well, why type annotations are necessary and this error is coming from a place where type annotations weren't explicitly provided, but should be.

Reply:
I took a stab at rewording this. I'm going to set the PR to auto-merge and we can rewrite this section of the docs in a future PR if you still think the follow up doesn't quite eliminate the weirdness you think it started out with.